cpp/src/datetime/datetime_ops.cu

```diff
@@ -110,6 +110,27 @@ struct extract_last_day_of_month {
   }
 };
 
+// Extract the day number of the year present in the timestamp
+struct extract_day_num_of_year {
```

Comment:
I think day of the year should use a simple lookup into a preintegrated array, not a while loop (remember, every thread will be doing this!

```

// defined globally (note not __constant__ since concurrent threads will be accessing different locations)
static __device__ const int16_t days_since_jan1[2][12] =
{
  { 0,31,59,90,120,151,181,212,243,273,304,334},
  { 0,31,60,91,121,152,182,213,244,274,305,335}
};

...
 
   template <typename Timestamp>
   CUDA_DEVICE_CALLABLE int16_t operator()(Timestamp const ts) const {
     using namespace simt::std::chrono;

     // Only has the days - time component is chopped off, which is what we want
     auto const days_since_epoch = floor<days>(ts);
     auto const date             = year_month_day(days_since_epoch);
     return days[date.year().is_leap()][date.month()-1] + date.day() - 1;
   }
```

Simpler and faster.

Reply:
ok - but, how much slower do you think looping in the worst case for ~ 12 times for each date (concurrently) would be? i did this only because the array was smaller and did not want to hard code those numbers. in any case, all this will probably go away sometime in favor of something along the lines of the following:

```
    const year_month_day ymd{floor<days>(ts)};
    const year_month_day pymd{--ymd.year(), December, day{31}};
    const auto dofyr = {sys_days{ymd} - sys_days{pymd}};
    return dofyr.count();
```

Reply:
We're talking thousands of threads, all looping a different number of iterations. That's a lot of parallel divergence.  Whereas with the __device__ array you have a tiny 48 byte array that every thread indexes into -- it will likely hit in L1 cache for most warps of every thread block.  

you can benchmark, but my money is on my suggested code. I wouldn't even be surprised if it is faster than your 4-line solution depending on the cost of `is_leap()`.  Might be simpler to  just get the year from the input timestamp, then construct Dec 31 the year before as a timestamp, subtract from the input timestamp...

Reply:
- all time in microsecs
- on v100

| num_timestamps | v0 - initial version | v1 - your version | v2 - 4line version |
| :---: |:---:| :---:| :---:|
| 1m | 96 | 32 | 33 |
| 10m | 763 | 193 | 201 |
| 25m | 1886 | 466 | 486 |
| 50m | 3756 | 919 | 968 |
| 100m | 7501 | 1822 | 1917 |

Reply:
I didn't see this comment, possibly because you resolved the discussion. If your code above is that close in performance without the __device__ array, feel free to switch to it.

BTW, this is why we ask that a google benchmark be  added for every new feature. (We haven't gotten to the point of enforcing that as a requirement yet...)